File: automerge-wasm/test/blame.ts
import { describe, it } from 'mocha';
//@ts-ignore
import assert from 'assert'
//@ts-ignore
import { BloomFilter } from './helpers/sync'
import { create, loadDoc, SyncState, Automerge, encodeChange, decodeChange, initSyncState, decodeSyncMessage, decodeSyncState, encodeSyncState, encodeSyncMessage } from '..'
import { DecodedSyncMessage, Hash } from '..'

describe('Automerge', () => {
  describe('blame', () => {
    it('should be able to blame text segments on change sets', () => {
      let doc1 = create()
      let text = doc1.make("_root", "notes","hello little world")
      let h1 = doc1.getHeads();

      let doc2 = doc1.fork();
      doc2.splice(text, 5, 7, " big");
      doc2.text(text)
      let h2 = doc2.getHeads();
      assert.deepEqual(doc2.text(text), "hello big world")

      let doc3 = doc1.fork();
      doc3.splice(text, 0, 0, "Well, ");
      let h3 = doc3.getHeads();
      assert.deepEqual(doc3.text(text), "Well, hello little world")

      doc1.merge(doc2)
      doc1.merge(doc3)
      assert.deepEqual(doc1.text(text), "Well, hello big world")
      let blame = doc1.blame(text, h1, [h2, h3])

      assert.deepEqual(blame, [
        { add: [ { start: 11, end: 15 } ], del: [ { pos: 15, val: ' little' } ] },
        { add: [ { start: 0,  end: 6  } ], del: [] }
      ])
    })
  })
})

File: automerge/tests/blame.rs
use automerge::transaction::Transactable;
use automerge::{AutoCommit, AutomergeError, ROOT};

/*
mod helpers;
use helpers::{
    pretty_print, realize, realize_obj,
    RealizedObject,
};
*/

#[test]
fn simple_blame_text() -> Result<(), AutomergeError> {
    let mut doc = AutoCommit::new();
    let note = doc.set_object(&ROOT, "note", automerge::ObjType::Text)?;
    doc.splice_text(&note, 0, 0, "hello little world")?;
    let baseline = doc.get_heads();
    assert!(doc.text(&note).unwrap() == "hello little world");
    let mut doc2 = doc.fork();
    doc2.splice_text(&note, 5, 7, " big")?;
    let h2 = doc2.get_heads();
    assert!(doc2.text(&note)? == "hello big world");
    let mut doc3 = doc.fork();
    doc3.splice_text(&note, 0, 0, "Well, ")?;
    let h3 = doc3.get_heads();
    assert!(doc3.text(&note)? == "Well, hello little world");
    doc.merge(&mut doc2)?;
    doc.merge(&mut doc3)?;
    let text = doc.text(&note)?;
    assert!(text == "Well, hello big world");
    let cs = vec![h2, h3];
    let blame = doc.blame(&note, &baseline, &cs)?;
    assert!(&text[blame[0].add[0].clone()] == " big");
    assert!(blame[0].del[0] == (15, " little".to_owned()));
    //println!("{:?} == {:?}", blame[0].del[0] , (15, " little".to_owned()));
    assert!(&text[blame[1].add[0].clone()] == "Well, ");
    //println!("- ------- blame = {:?}", blame);
    Ok(())
}
